NE-1744: Updating instructions for GettingStarted

Thealisyed opened the pull request.

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -204,8 +194,10 @@ $ oc new-app -n demo-app --name foo-app https://github.com/openshiftdemos/cakeph
-
+```
+```
 $ oc rollout status deployment -w -n demo-app foo-app
+oc rollout status deployment -w -n demo-app foo-app
````
Did you accidentally duplicate the line?

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -229,11 +221,7 @@ spec:
 ```
-Wait for the gateway deployment to be ready:
-```console
-$ oc wait -n openshift-ingress --for=condition=ready gateways.gateway.networking.k8s.io demo-gateway
````
Would it make sense to keep this but change "Ready" to "Programmed"?

candita replied:
+1

candita replied:
Was this change inadvertantly deleted?

Miciah reviewed (commented):
I think it could be confusing to update these instructions given that https://www.redhat.com/en/blog/gateway-api-with-openshift-networking-developer-preview-update still refers to OpenShift 4.13 and also references these instructions.  It would be safer to note the differences between OpenShift 4.13 and 4.18.

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -3,34 +3,24 @@
-  load balancers (e.g. GCP, AWS, and Azure)
-* If OSSM is already installed, it must be OSSM 2.4.0
+  load balancers (e.g. GCP, AWS and Azure)
+* If OSSM is already installed, it must be OSSM 2.6.2
```
```suggestion
* If OSSM is already installed, the minimum version must be OSSM 2.6.2
```

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -3,34 +3,24 @@
 * A non-production OpenShift Cluster, version 4.13.5 or greater, on a cloud platform that supports Kubernetes service
-  load balancers (e.g. GCP, AWS, and Azure)
-* If OSSM is already installed, it must be OSSM 2.4.0
+  load balancers (e.g. GCP, AWS and Azure)
```
We use the Oxford comma both by convention and by rule: https://github.com/openshift/enhancements/blob/master/guidelines/README.md?plain=1#L28
https://github.com/openshift/enhancements/blob/8d475dab4d176799814e1058264023a7f8762720/CONVENTIONS.md?plain=1#L58-L59
```suggestion
  load balancers (e.g. GCP, AWS, and Azure)
```

Thealisyed replied:
Ack, good to know, I was looking at this with the heathen comma in mind

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -3,34 +3,24 @@
 a simple HTTPRoute and backend to demo basic Gateway API functionality.
 
-Note: Since this is still a Developer Preview, it is important to note that it may not be flawless, and it is possible
+Note: Since this is still a Developer Preview, it is important to note that it may not be flawless and it is possible
```
This change is not grammatically correct.
```suggestion
Note: Since this is still a Developer Preview, it is important to note that it may not be flawless, and it is possible
```

Miciah commented:
/assign

candita commented:
/assign

Thealisyed commented:
Added an 'Instructions' section at the beginning of file outlining the key differences between 4.13 and 4.18. Reverted the changes of the rest of the file to reflect 4.13 as per Miciah's first comment

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -67,12 +80,11 @@ OSSM. Wait a minute or two for the SMCP to be created and become ready:
 
 OSSM will create two new Istio deployments in the `openshift-ingress` namespace. Ensure the pods for the deployments are running. These will need to become ready for the SMCP to also go ready:
 ```console
-$ oc get deployment -n openshift-ingress
````
I think you inadvertantly deleted this command.

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -204,7 +215,8 @@ $ oc new-app -n demo-app --name foo-app https://github.com/openshiftdemos/cakeph
      'oc expose service/foo-app'
     Run 'oc status' to view your app.
-
+```
````
This seems out of place.

candita commented:
> Added an 'Instructions' section at the beginning of file outlining the key differences between 4.13 and 4.18. Reverted the changes of the rest of the file to reflect 4.13 as per Miciah's first comment

I think he meant to keep both the old text and add the new text where they were, but point out the change, e.g.

> Prior to 4.18:
> oc wait -n openshift-ingress --for=condition=programmed gateways.gateway.networking.k8s.io demo-gateway-openshift-default

> 4.18 and beyond:
> oc wait -n openshift-ingress --for=condition=ready gateways.gateway.networking.k8s.io demo-gateway-openshift-default
>

lihongan commented:
Do we still need this PR? or shall we update the content to align with 4.19 ?

Thealisyed commented:
/remove-lifecycle stale 

need to priority this once i am back from PTO, should be a quick follow up commit

Thealisyed commented:
Updating the content to align with 4.20

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,55 +1,69 @@
-* If OSSM is already installed, you must not have any existing ServiceMeshControlPlane CRs
-* The [OpenShift CLI](https://docs.openshift.com/container-platform/4.13/cli_reference/openshift_cli/getting-started-cli.html#cli-installing-cli_cli-developer-commands) binary (oc)
+  * A non-production OpenShift Cluster, version 4.13.5 or greater, on a cloud platform that supports Kubernetes service load balancers (e.g. GCP, AWS, and Azure)
+  * If OSSM is already installed, it must be OSSM 2.4.0 for 4.13
```
```suggestion
  * If OSSM is already installed, it must be at least 
  ** OSSM 2.6 for OpenShift 4.18
  ** OSSM 3.0.1 for OpenShift 4.19
  ** OSSM 3.1.0 for OpenShift 4.20
```

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,55 +1,69 @@
-* The [OpenShift CLI](https://docs.openshift.com/container-platform/4.13/cli_reference/openshift_cli/getting-started-cli.html#cli-installing-cli_cli-developer-commands) binary (oc)
+  * A non-production OpenShift Cluster, version 4.13.5 or greater, on a cloud platform that supports Kubernetes service load balancers (e.g. GCP, AWS, and Azure)
+  * If OSSM is already installed, it must be OSSM 2.4.0 for 4.13
+    * Version Note: For OpenShift 4.20, the minimum required OSSM version is 2.6.2.
```
This can be removed.

candita replied:
delete this line

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,55 +1,69 @@
+  * If OSSM is already installed, it must be OSSM 2.4.0 for 4.13
+    * Version Note: For OpenShift 4.20, the minimum required OSSM version is 2.6.2.
+  * If OSSM is already installed, you must not have any existing ServiceMeshControlPlane CRs
+  * The [OpenShift CLI](https://docs.openshift.com/container-platform/4.13/cli_reference/openshift_cli/getting-started-cli.html#cli-installing-cli_cli-developer-commands) binary (`oc`)
```
Update to 4.20 link.

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,55 +1,69 @@
 
-Note: Since this is still a Developer Preview, it is important to note that it may not be flawless, and it is possible
-for bugs to be present.
+Note: Since this is still a Developer Preview, it is important to note that that it may not be flawless, and it is possible for bugs to be present.
```
Since we now have a GA, we'll need to reword alot of this.
```suggestion
Gateway API via Ingress Operator is GA in OpenShift 4.19.  These intstructions were written for versions prior 4.19, when Gateway API was considered a Developer Preview.  When using versions prior to 4.19, it is important to note that it is possible for bugs to be present.
```

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,55 +1,69 @@
-* If OSSM is already installed, it must be OSSM 2.4.0
-* If OSSM is already installed, you must not have any existing ServiceMeshControlPlane CRs
-* The [OpenShift CLI](https://docs.openshift.com/container-platform/4.13/cli_reference/openshift_cli/getting-started-cli.html#cli-installing-cli_cli-developer-commands) binary (oc)
+  * A non-production OpenShift Cluster, version 4.13.5 or greater, on a cloud platform that supports Kubernetes service load balancers (e.g. GCP, AWS, and Azure)
```
Let's pick a more recent version, like 4.18.0.
```suggestion
  * Prior to GA 4.19, use a non-production OpenShift Cluster
  * OpenShift version 4.18 or greater, on a cloud platform that supports Kubernetes service load balancers (e.g. GCP, AWS, and Azure)
```

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,55 +1,69 @@
+  * A non-production OpenShift Cluster, version 4.13.5 or greater, on a cloud platform that supports Kubernetes service load balancers (e.g. GCP, AWS, and Azure)
+  * If OSSM is already installed, it must be OSSM 2.4.0 for 4.13
+    * Version Note: For OpenShift 4.20, the minimum required OSSM version is 2.6.2.
+  * If OSSM is already installed, you must not have any existing ServiceMeshControlPlane CRs
```
```suggestion
  * In OpenShift 4.18, if OSSM is already installed, you must not have any existing ServiceMeshControlPlane CRs
```

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,55 +1,69 @@
-2. Enable the feature gate
-3. Create a GatewayClass with the controller name “openshift.io/gateway-controller”
+
+1.  Grant the Ingress Operator the cluster-admin role (no longer required in 4.20)
+2.  Enable the feature gate (no longer required to 4.20)
+3.  Create a GatewayClass with the controller name `openshift.io/gateway-controller`
```
```suggestion
1.  For OCP versions before 4.20, grant the Ingress Operator the cluster-admin role, enable the feature gate, and
create a GatewayClass with the controller name `openshift.io/gateway-controller`
2. For OCP 4.20 and later, just create a GatewayClass with the controller name `openshift.io/gateway-controller/v1` (note the additional version suffix)
```

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -229,22 +252,32 @@ spec:
 
-Thanks for trying OpenShift Istio Gateway API. If you'd like to learn more, go to section on [Getting Creative](./GettingCreative.md).
\ No newline at end of file
+Thanks for trying OpenShift Istio Gateway API. If you'd like to learn more, go to section on [Getting Creative](https://www.google.com/search?q=./GettingCreative.md).
```
It shouldn't need the new url.  If it does, use the correct github.com url.

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -1,55 +1,69 @@
+
 ```console
 $ oc adm policy add-cluster-role-to-user cluster-admin -z ingress-operator -n openshift-ingress-operator
 clusterrole.rbac.authorization.k8s.io/cluster-admin added: "ingress-operator"
 ```
-It is anticipated that this need for cluster-admin permissions will be eliminated in a future release of Gateway API in
-OpenShift.
+
+It is anticipated that this need for cluster-admin permissions will be eliminated in a future release of Gateway API in OpenShift.
+
+### 2\. Enable the Feature Gate
 
 Enable the feature gate for Gateway API. This will instruct the cluster Ingress Operator to install the Gateway API CRDs.
-Please note that, since feature gates are stored in the machine config, patching the feature gate config will result in
-the machine config operator rebooting each node sequentially, according to your machine config pool configuration:
+Please note that, since feature gates are stored in the machine config, patching the feature gate config will result in the machine config operator rebooting each node sequentially, according to your machine config pool configuration:
+* **OpenShift 4.20 and beyond:** This step is no longer required and should be skipped.
+
 ```console
 $ oc patch featuregates/cluster --type=merge --patch='{"spec":{"featureSet":"CustomNoUpgrade","customNoUpgrade":{"enabled":["GatewayAPI"]}}}'
 featuregate.config.openshift.io/cluster patched
 ```
 
 Wait for the ingress operator to successfully install the Gateway API CRDs:
+
 ```console
 $ oc get crd gatewayclasses.gateway.networking.k8s.io httproutes.gateway.networking.k8s.io gateways.gateway.networking.k8s.io referencegrants.gateway.networking.k8s.io
-NAME                                       CREATED AT
-gatewayclasses.gateway.networking.k8s.io   2023-04-04T18:02:55
-httproutes.gateway.networking.k8s.io        2023-04-04T18:02:55Z
-gateways.gateway.networking.k8s.io          2023-04-04T18:02:55Z
-referencegrants.gateway.networking.k8s.io   2023-04-04T18:02:56Z
+NAME                                        CREATED AT
+gatewayclasses.gateway.networking.k8s.io    2025-02-04T10:49:54Z
+httproutes.gateway.networking.k8s.io        2025-02-04T10:49:55Z
+gateways.gateway.networking.k8s.io          2025-02-04T10:49:54Z
````
This whole section should be re-written with a Header like: "For OCP version before 4.20".

candita commented:
Hey @Thealisyed please take a fresh look at this and update it for 4.18+.

lihongan commented:
cc @rhamini3

rhamini3 commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,55 +1,69 @@
+
+First, grant the Ingress Operator the cluster-admin role. This role is required in order for the Ingress Operator to create the SMCP CR for OSSM:
+
+  * **OpenShift 4.20 and beyond:** This step is no longer required and should be skipped.
```
do we even provide support for OSSM 2.x? do we need this section about SMCP?

candita replied:
This is an old doc being updated for 4.20+, and we still allow people to experiment on previous versions.

rhamini3 commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,55 +1,69 @@
+
+It is anticipated that this need for cluster-admin permissions will be eliminated in a future release of Gateway API in OpenShift.
+
+### 2\. Enable the Feature Gate
```
need to remove feature gate stuff

candita replied:
This is still valid for 4.18.

rhamini3 commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,55 +1,69 @@
-Please note that, since feature gates are stored in the machine config, patching the feature gate config will result in
-the machine config operator rebooting each node sequentially, according to your machine config pool configuration:
+Please note that, since feature gates are stored in the machine config, patching the feature gate config will result in the machine config operator rebooting each node sequentially, according to your machine config pool configuration:
+* **OpenShift 4.20 and beyond:** This step is no longer required and should be skipped.
```
this should be 4.19 and above? It went GA then

rhamini3 commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -62,30 +76,34 @@ EOF
 gatewayclass.gateway.networking.k8s.io/openshift-default created
```
for us we always use the controller name as `controllerName: openshift.io/gateway-controller/v1`

candita replied:
This has to be updated to distinguish between old and new controllerNames.

rhamini3 commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -62,30 +76,34 @@ EOF
+openshift-gateway   5/5     ComponentsReady   ["default"]   2.4.0     10m
 ```
 
+**Version Note:** For OpenShift 4.20, the SMCP version will be 2.6.2 or greater.
````
if we create gatewayclass now, doesnt it automatically install OSSM 3?

rhamini3 commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -101,9 +119,8 @@ spec:
-match a subdomain of the cluster domain. The Gateway must be created in the `openshift-ingress` namespace and specify
-the GatewayClass that we created earlier in the `spec.gatewayClassName` field.
+First, create the Gateway with a listener (`spec.listeners` instance) that will listen on a port for HTTP requests that match a subdomain of the cluster domain. The Gateway must be created in the `openshift-ingress` namespace and specify the GatewayClass that we created earlier in the `spec.gatewayClassName` field.
+
```
It might be worthwhile to mention that the gateway can be created in another namespace, but just wont have DNSRecords

candita replied:
I disagree.  Instead, let's keep it simple for those just getting started.

rhamini3 commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,91 +1,141 @@
+gatewayclasses.gateway.networking.k8s.io    2025-02-04T10:49:54Z
+httproutes.gateway.networking.k8s.io        2025-02-04T10:49:55Z
+gateways.gateway.networking.k8s.io          2025-02-04T10:49:54Z
+referencegrants.gateway.networking.k8s.io   2025-02-04T10:49:55Z
```
4.20 beyond also supports grpc route crds, do we want to add that in?

candita replied:
It was added.

rhamini3 commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -1,91 +1,141 @@
+    $ oc get istio -A
+    NAME                NAMESPACE           PROFILE   REVISIONS   READY   IN USE   ACTIVE REVISION     STATUS    VERSION   AGE
+    openshift-gateway   openshift-ingress             1           1       0        openshift-gateway   Healthy   v1.26.2   2m
+    ```
````
isn't istio cluster scoped? `-A` should not be required

rhamini3 commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -1,91 +1,141 @@
+    router-default             2/2     2            2           6h4m
+    ```
 
 Gateway API and OSSM have been installed successfully and are ready to use. We are now ready to create Gateways and HTTPRoutes.
````
would it be a good idea to mention that at this point when all the deployments are successful, the gatewayClass will then show `Accepted` status?

rikatz commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,91 +1,141 @@
+    * OSSM 3.0.1 for OpenShift 4.19
+    * OSSM 3.1.0 for OpenShift 4.20
+  * In OpenShift 4.18, if OSSM is already installed, you must not have any existing ServiceMeshControlPlane CRs
+  * The [OpenShift CLI](https://docs.okd.io/latest/cli_reference/openshift_cli/getting-started-cli.html#:~:text=using%20an%20RPM.-,Installing%20the%20OpenShift%20CLI,-You%20can%20install) binary (`oc`)
```
shouldn't this point to official docs?

rikatz commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,91 +1,141 @@
-2. Enable the feature gate
-3. Create a GatewayClass with the controller name “openshift.io/gateway-controller”
+
+1.  For OCP versions before 4.20, grant the Ingress Operator the cluster-admin role, enable the feature gate, and
```
which feature gate? should it be part of the text here?

rikatz replied:
ah I can see it below. Does it make sense to remove these bullet points here and just direct users to the details then?

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,91 +1,140 @@
-It is anticipated that this need for cluster-admin permissions will be eliminated in a future release of Gateway API in
-OpenShift.
+
+It is anticipated that this need for cluster-admin permissions will be eliminated in a future release of Gateway API in OpenShift.
```
This could be removed, since we already say 

```
OpenShift 4.20 and beyond: This step is no longer required and should be skipped.
```

Miciah replied:
The redundant sentence is still present.

candita commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -1,91 +1,140 @@
+    gatewayclass.gateway.networking.k8s.io/openshift-default created
+    ```
+
+Behind the scenes, the Ingress Operator creates a ServiceMeshControlPlane (SMCP) CR, which is the configuration API for OSSM. Wait a minute or two for the SMCP to be created and become ready:
````
This is not quite correct, but it is corrected later in the text.
```suggestion
Behind the scenes, the Ingress Operator creates either an Istio or ServiceMeshControlPlane (SMCP) CR.  The SMCP CR is the configuration API for OSSM prior to OCP 4.20, and the Istio CR takes its place in 4.20 and later. Wait a minute or two for the Istio/SMCP to be created and become ready:
```

candita commented:
There are still a few nits in here, but I think we need to fast-track this before anything else changes.  Thanks @Thealisyed !

/lgtm
/approve

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,91 +1,140 @@
-for bugs to be present.
+In this section, we will show you how to install and configure Gateway API via the Ingress Operator. We will configure a simple HTTPRoute and backend to demo basic Gateway API functionality.
 
+Note: Gateway API via Ingress Operator is GA in OpenShift 4.19.  These instructions were written for versions prior to 4.19, when Gateway API was considered a Developer Preview.  When using versions prior to 4.19, it is important to note that bugs may be present.
```
This is a confusing statement, especially "bugs may be present".  The examples should have been tested on the Developer Preview versions, and they should still work on the GA version.  If there are any discrepancies, those should be noted in the document (for example, the example GatewayClass should have a note that the Developer Preview and GA use different controller names), or if that is too much effort, just say that the instructions were tested on the Developer Preview and might require some adjustment for GA.

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,91 +1,140 @@
-* The [OpenShift CLI](https://docs.openshift.com/container-platform/4.13/cli_reference/openshift_cli/getting-started-cli.html#cli-installing-cli_cli-developer-commands) binary (oc)
+  * Prior to GA 4.19, use a non-production OpenShift Cluster
+  * OpenShift version 4.18 or greater, on a cloud platform that supports Kubernetes service load balancers (e.g. GCP, AWS, and Azure)
+  * If OSSM is already installed, the minimum version must be:
```
Is it a _minimum_ version, or does the version need to be _exactly_ the one listed here?

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -1,91 +1,140 @@
+    openshift-gateway   5/5     ComponentsReady   ["default"]   2.4.0     10m
+    ```
+
+  * **OpenShift 4.20 and beyond:** OSSM 3.x uses Istio resources instead of SMCP.
````
```suggestion
  * **OpenShift 4.20 and beyond:** OSSM 3._y_ uses Istio resources instead of SMCP.
```
or simply
```suggestion
  * **OpenShift 4.20 and beyond:** OSSM 3 uses Istio resources instead of SMCP.
```

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -229,22 +287,43 @@ spec:
-gateway.gateway.networking.k8s.io/demo-gateway condition met
-```
+
+  * **Prior to OpenShift 4.20:** The condition is `ready`.
````
```suggestion
  * **Prior to OpenShift 4.20:** The condition is `Ready`.
```

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -229,22 +287,43 @@ spec:
+    ```console
+    $ oc wait -n openshift-ingress --for=condition=ready gateways.gateway.networking.k8s.io demo-gateway
+    ```
+  * **OpenShift 4.20 and beyond:** The condition is `programmed`.
````
```suggestion  
  * **OpenShift 4.20 and beyond:** The condition is `Programmed`.
```

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -229,22 +287,43 @@ spec:
+
+You should see a 200 response code. The response will include headers from the application:
+
+  * **Prior to OpenShift 4.20:** Response includes `istio-envoy` server header.
```
```suggestion
  * **Prior to OpenShift 4.19:** Response includes `istio-envoy` server header.
```
OpenShift 4.19 configures Istio not to set this header: https://github.com/openshift/cluster-ingress-operator/blob/release-4.19/pkg/operator/controller/gatewayclass/istio.go#L188-L191

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -229,22 +287,43 @@ spec:
+    ...
+    ```
+
+  * **OpenShift 4.20 and beyond:** Response includes application server header and `app: foo`.
````
`app: foo` is in the previous example for OpenShift 4.19 and earlier too, and I believe it is set by the application, not Envoy.

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,91 +1,140 @@
+To install Gateway API using the Ingress Operator:
+
+* **OCP 4.20 and later:** Create a GatewayClass with controller name `openshift.io/gateway-controller/v1` (see [Step 3](#3-create-the-gatewayclass))
+* **OCP versions before 4.20:** Follow Steps 1-3 below (grant cluster-admin role, enable the `GatewayAPI` feature gate, create GatewayClass)
```
OpenShift 4.19 has the controller name with the "/v1" suffix: https://github.com/openshift/cluster-ingress-operator/blob/release-4.19/pkg/operator/controller/names.go#L65

I think you need to fix a number of places that say "4.20 or later" and "before 4.20" to say "4.19 or later" and "before 4.19" instead.

candita commented:
/lgtm

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,91 +1,140 @@
-* The [OpenShift CLI](https://docs.openshift.com/container-platform/4.13/cli_reference/openshift_cli/getting-started-cli.html#cli-installing-cli_cli-developer-commands) binary (oc)
+  * Prior to GA 4.19, use a non-production OpenShift Cluster
+  * OpenShift version 4.18 or greater, on a cloud platform that supports Kubernetes service load balancers (e.g. GCP, AWS, and Azure)
+  * If OSSM is already installed, it must be at least:
```
I'm still not sure that "at least" is correct.

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,91 +1,140 @@
 
 Wait for the ingress operator to successfully install the Gateway API CRDs:
+
+> **Note:** OCP 4.20 and later also includes the `grpcroutes.gateway.networking.k8s.io` CRD for routing gRPC traffic.
```
```suggestion
> **Note:** OCP 4.19 and later also includes the `grpcroutes.gateway.networking.k8s.io` CRD for routing gRPC traffic.
```

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,91 +1,140 @@
+
+Behind the scenes, the Ingress Operator creates a ServiceMeshControlPlane (SMCP) CR, which is the configuration API for OSSM. Wait a minute or two for the SMCP to be created and become ready:
+
+  * **Prior to OpenShift 4.20:** OSSM uses ServiceMeshControlPlane (SMCP) resources.
```
```suggestion
  * **Prior to OpenShift 4.19:** OSSM uses ServiceMeshControlPlane (SMCP) resources.
```

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -1,91 +1,140 @@
+    gatewayclass.gateway.networking.k8s.io/openshift-default created
+    ```
+
+Behind the scenes, the Ingress Operator creates a ServiceMeshControlPlane (SMCP) CR, which is the configuration API for OSSM. Wait a minute or two for the SMCP to be created and become ready:
````
This statement about the SMCP belongs under the "Prior to OpenShift 4.19" section.

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -1,91 +1,140 @@
+    openshift-gateway   5/5     ComponentsReady   ["default"]   2.4.0     10m
+    ```
+
+  * **OpenShift 4.20 and beyond:** OSSM 3 uses Istio resources instead of SMCP.
````
```suggestion
  * **OpenShift 4.19 and beyond:** OSSM 3 uses Istio resources instead of SMCP.
```

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -1,91 +1,140 @@
+
+OSSM will create Istio deployments in the `openshift-ingress` namespace. Ensure the pods for the deployments are running.
+
+  * **Prior to OpenShift 4.20:** OSSM creates a shared `istio-ingressgateway` deployment and the control plane.
```
```suggestion
  * **Prior to OpenShift 4.19:** OSSM creates a shared `istio-ingressgateway` deployment and the control plane.
````

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -1,91 +1,140 @@
+    router-default            2/2     2            2           6h4m
+    ```
+
+  * **OpenShift 4.20 and beyond:** Only the Istio control plane (`istiod`) is created at this stage. Individual gateway deployments are created when you create Gateway resources.
````
```suggestion
  * **OpenShift 4.19 and beyond:** Only the Istio control plane (`istiod`) is created at this stage. Individual gateway deployments are created when you create Gateway resources.
````

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
```diff
@@ -127,24 +175,31 @@ gateway.gateway.networking.k8s.io/demo-gateway created
 
 By default, Istio will [automatically provision](https://istio.io/latest/docs/tasks/traffic-management/ingress/gateway-api/#automated-deployment) a gateway deployment and service with the same name upon creation of this Gateway object:
+
+* **OpenShift 4.20 and beyond:** The deployment name changes to `demo-gateway-openshift-default`.
```
```suggestion
* **OpenShift 4.19 and beyond:** The deployment name changes to `demo-gateway-openshift-default`.
```

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -229,22 +287,43 @@ spec:
-gateway.gateway.networking.k8s.io/demo-gateway condition met
-```
+
+  * **Prior to OpenShift 4.20:** The condition is `Ready`.
````
Given my previous comments, I suspect this should be "4.19", not "4.20".  Please double-check.

Miciah commented on docs/blogs/EnhancedDevPreviewGatewayAPI/GettingStarted.md:
````diff
@@ -229,22 +287,43 @@ spec:
+    ```console
+    $ oc wait -n openshift-ingress --for=condition=ready gateways.gateway.networking.k8s.io demo-gateway
+    ```
+  * **OpenShift 4.20 and beyond:** The condition is `Programmed`.
````
Likewise, please double-check that "4.20" is correct.

Miciah commented:
/approve

@candita and I agreed that we will merge this PR and follow up with corrections in a separate PR.

Miciah commented:
Ah, I forgot—there's no automation on this repository, so I need to go through the drudgery of manually clicking the green button.  Such toil!
